src/GraphQL/Execution/ExecutionHelper.cs

```diff
@@ -52,20 +54,60 @@ public static class ExecutionHelper
 
             foreach (var arg in definitionArguments.List!)
             {
-                var value = astArguments?.ValueFor(arg.Name);
+                GraphQLArgument? argNode = null;
+                if (astArguments != null)
+                {
+                    foreach (var node in astArguments)
+                    {
+                        if (node.Name.Value.Equals(arg.Name))
+                        {
+                            argNode = node;
+                        }
+                    }
+                }
+                var value = argNode?.Value;
                 var type = arg.ResolvedType!;
 
-                values[arg.Name] = CoerceValue(type, value, variables, arg.DefaultValue);
+                values[arg.Name] = CoerceValue(type, value, new CoerceValueContext
+                {
+                    Argument = argNode,
+                    Document = document,
+                    Directive = directive,
+                    ParentNode = fieldOrFragmentSpread,
+                    Variables = variables,
+                }, arg.DefaultValue);
             }
 
             return values;
         }
 
+        internal readonly ref struct CoerceValueContext
```

Comment:
why ref ?

Reply:
More efficient

Reply:
Is it noticeable here in practice or made for general reasons?

Reply:
ref struct only needs to pass a single pointer to the stack where the struct is stored. Readonly ensures it cannot be changed. Struct would need to assign/copy all variables for each call. So for any readonly struct, readonly ref struct should always be faster (so long as it does not need to be copied to the heap). More so the larger the struct. I’m sure it makes little difference in practice, especially here. It’s internal for now; perhaps if it was public then we would use a regular struct or a class. We also could not use init properties if it was public